README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# habCluster

<!-- badges: start -->
<!-- badges: end -->

Based on landscape connectivity, use Community Detection Algorithm to
find structure of raster and return a polygon representing the boundary
of the clusters.

## Installation

You can install the development version of habCluster from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("qiangxyz/habCluster")
```

## Example

This is a basic example which shows you how to find the cluster of
lands:

``` r
library(sf)
library(raster)
library(habCluster)
```

Read in habitat suitability index (HSI) data of wolf in Europe. The HSI
values of the cells in the raster indicate how smoothly the wolfs can
moved in the cells, and can be used to represent the connection between
cells as habitat.

``` r
hsi.file = system.file("extdata","wolf3_int.tif",package="habCluster")
wolf = raster(hsi.file)
```

Find habitat cluster using Leiden Algorithm. Raster for habitat
suitability will be resampled to 40 km (40000 m), to reduce calculation
amount. Set cluster_resolution_parameter to 0.02 to control the cluster
size. Note that the parameter of cellsize controls the spatial scale
analysis is performed, while the parameter of rp is used to control
cluster size.

``` r
clst = cluster(wolf, method = cluster_leiden, cellsize = 40000, rp = 0.02, silent = FALSE)
#> 
#> resampling...
#> extracting edges...
#> create graph...
#> finding clusters...
#> preparing results...
```

We can also embed plots, for example:

``` r
image(wolf, col = terrain.colors(100,rev = T), asp = 1)
boundary = clst$boundary
plot( boundary$geometry, add=TRUE, asp=1, border = "lightseagreen")
```

<img src="man/figures/README-cluster-1.png" width="100%" />

Or, we can discard small patches before plotting:

``` r
image(wolf, col = terrain.colors(100,rev = T), asp = 1)
boundary$area = as.numeric(st_area(boundary))
boundary = boundary[boundary$area > 40000*40000,]
plot( boundary$geometry, add=TRUE, asp=1, border = "lightseagreen")
```

<img src="man/figures/README-discard-1.png" width="100%" />

Using stars to manipulate the raster will make the computation more
quick:

``` r
hsi.file = system.file("extdata","wolf3_int.tif",package="habCluster")
wolf = read_stars(hsi.file)
clst = cluster(wolf, method = cluster_leiden, cellsize = 40000, rp = 0.02, silent = FALSE)
#> 
#> resampling...
#> extracting edges...
#> create graph...
#> finding clusters...
#> preparing results...
```

## How to Cite

Zhang,C., J. Li, B. Yang, Q. Dai\*. (2022). habCluster: Identifying
Geographical Boundary among Intraspecific Units Using Community
Detection Algorithm in R. bioRxiv 2022.05.06.490926.


NEWS.md
# habCluster 1.0.4

# habCluster 1.0.3
## Minor Bug Fixes and Improvements
* fix error in references describing the methods in DESCRIPTION.

## Changes
* small change on the the sample data.
* change the default method for function cluster from cluster_louvain to cluster_fast_greedy

# habCluster 1.0.2
## Minor Bug Fixes and Improvements
* add references describing the methods in DESCRIPTION.
* fixed the "dontrun" fails to comply with coding conventions.
* fixed the 'Suggests' in DESCRIPTION that fails to comply with coding conventions.


# habCluster 1.0.1
## Enhancements
* cluster() and raster2graph accept both stars and RasterLayer as parameter now.

## Changes
* cluster() returns an object of sf, instead of SpatialPolygonsDataFrame.


# habCluster 1.0.0
* First public release.


tests/testthat/test-cluster.R

test_that("cluster method", {
  library(habCluster)
  require(stars)
  require(sf)
  # read data
  hsi.file = system.file("extdata","wolf3_int.tif",package="habCluster")
  wolf = read_stars(hsi.file)

  clst = cluster(wolf, method = cluster_leiden, cellsize = 80000, rp = 0.02)

  # check results are right objects
  testthat::expect_s3_class(clst$boundary,"sf")
  testthat::expect_s3_class(clst$communities ,"communities")

})


tests/testthat/test-raster2Graph.R

test_that("raster2Graph method", {
  require(raster)
  # read data
  hsi.file = system.file("extdata","wolf3_int.tif",package="habCluster")
  wolf = raster(hsi.file)

  g = raster2Graph(wolf, 80000)

  # check results are right objects
  testthat::expect_s4_class(g$raster,"RasterLayer")
  testthat::expect_s3_class(g$graph ,"igraph")
})


src/getEdge.cpp
#include <Rcpp.h>
using namespace Rcpp;
using namespace std;


long getCellID(int &r, int &c, int &nrow){
  return nrow*(c)+r+1;
}

bool hasEdge(int &from, int &to, map<int,set<int>> &mp){
  map<int,set<int>>::iterator it=mp.find(from);
  if(it==mp.end()){
    return false;
  }else{
    return it->second.find(to) != it->second.end();
  }
}

// [[Rcpp::export]]
DataFrame getEdgeDF(NumericMatrix &m){
  map <int, set<int>> oldEdges;
  vector<int> fromV;
  vector<int> toV;
  vector<double>weightV;

  int nrow = m.rows();
  int ncol = m.cols();

  for(int r0 = 0; r0 < nrow; r0++){
    for(int c0 = 0; c0 < ncol; c0++){
      if(NumericVector::is_na(m(r0,c0))){
        continue;
      }

      for(int r1 = r0-1; r1<=r0+1; r1++){
        if(r1 < 0 || r1 == nrow){
          continue;
        }

        for(int c1 = c0-1; c1<c0+1; c1++){
          if(NumericVector::is_na(m(r1,c1))){
            continue;
          }
          if(c1< 0 || c1 == ncol){
            continue;
          }
          if(c0==c1&&r0==r1){
            continue;
          }

          int from = getCellID(r0,c0,nrow);
          // cout<< r0 <<","<< c0 <<": "<<from<<endl;
          int to = getCellID(r1,c1,nrow);

          // check whether the edge is duplicated as undirected;
          if(hasEdge(to,from,oldEdges)){
            continue;
          }


          double w = 0.0;

          if (r0 == r1 || c0 == c1) {
            w = 0.5 * (m(r0,c0) + m(r1,c1));
          } else{
            w =  0.7071068 * (m(r0,c0) + m(r1,c1));
          }

          oldEdges[from].insert(to);
          fromV.push_back(from);
          toV.push_back(to);
          weightV.push_back(w);
        }

      }
    }
  }

  DataFrame df = DataFrame::create( Named("v1") = fromV,
                                    Named("v2") = toV,
                                    Named("weight") = weightV);
  return df;
}


R/raster2Graph.R

#' Create a graph from an raster according the connection between cells
#'
#' @param r  An object of stars or RasterLayer. The value of each cell of the raster is the ‘smoothness’ to indicate how easy the cell connecting with neighbor cells.
#' @param cellsize Numeric. Re-sample the input raster to given resolution and use the re-sampled raster to build graph. Set this to NULL if using the original resolution of of the input raster.
#' @param relative.distance Boolean. If FALSE, absolute distance between cells is used to compute the edge weight; otherwise, relative distance between cells is used. Default is TRUE.
#' @param silent Boolean. A logical indicating if some “progress report” should be given. Default is TRUE.
#'
#' @return a list with an graph and the re-sampled raster (a object of stars). The graph is igraph object, with cells as node and connections as weight.
#' @export
#'
#' @examples
#' # read in habitat suitability data of wolf in Europe
#' library(stars)
#' hsi.file = system.file("extdata","wolf3_int.tif",package="habCluster")
#' wolf = read_stars(hsi.file)
#' # build graph from raster
#' g = raster2Graph(wolf, 40000)

raster2Graph  <- function(r, cellsize=NULL, relative.distance = TRUE, silent=TRUE){

  if(is(r,"RasterLayer")){
    r = stars::st_as_stars(r)
  }

  r2 = NULL
  if(!is.null(cellsize)){
    if(!silent){
    cat('\nresampling...')
    }
    r2 = stars::st_warp(r, crs=sf::st_crs(r), cellsize = cellsize)
  }else{
    r2 = r
  }

  matrix = r2[[1]]
  if(!silent){
  cat('\nextracting edges...')
  }
  edf = getEdgeDF(matrix)
  if(relative.distance == FALSE){
    dim = stars::st_dimensions(r2)
    res.x = abs(dim$x$delta)
    res.y = abs(dim$y$delta)
    if(res.x != res.y){
      stop("Cell size in the x and y dimension are not equal.")
    }

    edf$weight = edf$weight * res.x
  }
  if(!silent){
  cat('\ncreate graph...')
  }
  g =list()
  g$graph = igraph::graph_from_data_frame(edf, directed=F)
  g$raster = r2
  return(g)
}



R/habCluster-package.R
#' @keywords internal
"_PACKAGE"

#' The habCluster Package: Detecting Spatial Clustering Based on Connection Cost between Grids
#'
#' @description
#' Spatial explicitly identify the structure of landscape base on cost (weighted) connection at grid scale, using community detection algorithm from graph theory.
#'
#' @details Based on landscape connectivity, spatial boundaries were identified using community detection algorithm at grid level. Methods using raster as input and the value of each cell of the raster is the ‘smoothness’ to indicate how easy the cell connecting with neighbor cells. The method can be use to find habitat clusters for wildlife from a map of habitat suitability index, and thus can be used to detect the spatial boundaries among populations or other intraspecific units.
#'

#'
#' Key content of the 'envi' package include:\cr
#'
#' \bold{Spatial Cluster Detection}
#'
#' \code{\link{cluster}} Clustering cells from a raster by Community Detection Algorithm according to the connections between them and return a cluster map.
#'
#' @name habCluster-package
#' @aliases habCluster-package habCluster
#' @docType package
#'
#' @section Dependencies: The 'habCluster' package relies heavily upon \code{\link{raster}}, \code{\link{Rcpp}}, and \code{\link{igraph}}.
#'
#' @author Qiang Dai\cr \emph{Chengdu Institute of Biology, Chinese Academy of Sciences.}\cr
#'
#' Maintainer: Qiang Dai \email{daiqiang@@cib.ac.cn}
#'
#' @keywords package
#'
#'
## usethis namespace: start
## usethis namespace: end
#' @useDynLib habCluster
#' @importFrom Rcpp sourceCpp
#' @importFrom igraph membership cluster_fast_greedy cluster_louvain cluster_optimal cluster_walktrap cluster_infomap
#' @importFrom raster raster rasterToPolygons
NULL



R/cluster.R


#' Clustering cells from a raster by Community Detection Algorithm according to the connections between them and return a cluster map
#'
#' @description This function use Community Detection Algorithm to find structure of raster and return a polygon representing the boundary of the clusters.
#'
#' @param r An object of stars or RasterLayer. The value of each cell of the raster is the ‘smoothness’ to indicate how easy the cell connecting with neighbor cells.
#' @param method method from package igraph used to finding community structure. (see details below).
#' @param cellsize Numeric. Re-sample the input raster to given resolution and use the resampled raster to find community structure. Set this to NULL if using the original resolution of of the input raster,given the parameter r is an object of raster.
#' @param silent Boolean. A logical indicating if some “progress report” should be given. Default is TRUE.
#' @param ... Optional arguments to method. For example, can set resolution_parameter for cluster_leiden, or resolution for cluster_louvain.(see details below).
#'
#' @details Choice of the method used to finding community structure(see Mukerjee, 2021). The default method is cluster_fast_greedy, but could also be methods like cluster_leiden, cluster_walktrap, or cluster_louvain. If cluster_leiden is chosen, then we can use resolution_parameter to control the size of clusters. Higher resolution_parameter lead to more smaller clusters, while lower resolution_parameter lead to fewer larger clusters. The parameter of resolution for cluster_louvain is similar. More details about those methods can be found in the document for package "igraph".
#' @return A polygon of sf object for boundaries of habitat clusters,
#' and an object of communities defined in package igraph.
#'
#' @references
#' Mukerjee, S. (2021). A systematic comparison of community detection algorithms for measuring selective exposure in co-exposure networks. Scientific reports 11, 15218. https://doi.org/10.1038/s41598-021-94724-1\cr
#' Traag, V. A., Waltman, L., & van Eck, N. J. (2019). From Louvain to Leiden: guaranteeing well-connected communities. Scientific reports, 9(1), 5233. doi: 10.1038/s41598-019-41695-z\cr
#'
#' @export
#'
#' @examples
#' library(sf)
#' library(stars)
#'
#' # read in habitat suitability data of wolf in Europe
#' library(stars)
#' hsi.file = system.file("extdata","wolf3_int.tif",package="habCluster")
#' wolf = read_stars(hsi.file)
#'
#' # rescale raster value to 0 - 1
#' wolf = wolf / 100
#'
#' # find habitat cluster using Fast Greedy Algorithm.
#' # Raster will be resampled to 40 km, to cluser at the scale of 40 km and reduce calculation amount.
#' clst = cluster(wolf, method = cluster_fast_greedy, cellsize = 40000)
#'
#' # plot the results
#' image(wolf,col=terrain.colors(100,rev = TRUE),asp = 1)
#' boundary = clst$boundary
#' plot( boundary$geometry, add=TRUE, asp=1, border = "lightseagreen")
#'
#' # discard patches smaller than 1600 sqkm
#' boundary$area = as.numeric(st_area(boundary))
#' boundary = boundary[boundary$area > 40000*40000,]
#' image(wolf,col=terrain.colors(100,rev = TRUE),asp = 1)
#' plot( boundary$geometry, add=TRUE, asp=1, border = "lightseagreen")
#'
#' # can also use RasterLayer object#
#' library(raster)
#' wolf = read_stars(hsi.file)
#' wolf = wolf / 100
#' clst = cluster(wolf, method = cluster_leiden, cellsize = 40000, resolution_parameter = 0.0002)


cluster <- function(r=NULL, method=igraph::cluster_fast_greedy, cellsize=NULL, silent=TRUE,...){

  g = raster2Graph(r, cellsize, silent)

  if(!silent){
    cat('\nfinding clusters...')
  }

  if(identical(method, igraph::cluster_leiden)){
    clusters = igraph::cluster_leiden(g$graph,...)
  }else{
    clusters = method(g$graph,...)
  }
  if(!silent){
  cat('\npreparing results...')
  }

  cluster.id = igraph::membership(clusters)
  cell.id = as.integer(names(cluster.id))
  cluster.raster = g$raster
  cluster.raster[[1]][cell.id] = cluster.id
  boundaries = sf::st_as_sf(cluster.raster,merge=TRUE)
  names(boundaries)[1] = 'cluster.id'
  out=list()
  out$boundary = boundaries
  out$communities = clusters
  return(out)
}
